### ISSUE TITLE
[Feature Request]: logtail in push mode

### ISSUE THREAD

**Issue by XuPeng-SH** [matrixorigin/matrixone#6894]
### Is there an existing issue for the same feature request?

- [X] I have checked the existing issues.

### Describe the feature you'd like

Add support for push mode log shipping from DN to CN

**Comment by cnutshell** [comment]
Ask @aptend for the necessary context information firstly. Then make a design on the demand of computing group.

**Comment by cnutshell** [comment]
Design doc was reviewed, I'll construct prototype next.

**Comment by cnutshell** [comment]
Task is delayed. There are some complexity that need to be tackled:

**Comment by cnutshell** [comment]
Now PR #7363 should be merged if all reviewers approve that PR.
After merged, guys from CN and DN group could integrate `logtail push model` into mo based on PR #7363.

**Comment by cnutshell** [comment]
Now PR #7447 should be merged if all reviewers approve that PR.
Before guys from CN group could make request with `logtail push server`, there maybe two additional PR needed:
- Heartbeat `logtail server` address which need update proto buffer definition
- Split large `logtail response` if necessary

**Comment by cnutshell** [comment]
`logtail server address` is included in heartbeat after PR #7513.

> Now PR #7447 should be merged if all reviewers approve that PR. Before guys from CN group could make request with `logtail push server`, there maybe two additional PR needed:
> 
> * Heartbeat `logtail server` address which need update proto buffer definition
> * Split large `logtail response` if necessary
> 
> But these two additional PR are mainly related with integration test. It wouldn't block CN's development.

**Comment by m-schen** [comment]
start to write CN code today.

**Comment by cnutshell** [comment]
`Logtail response segmentation` is included in PR 7583.

> Now PR #7447 should be merged if all reviewers approve that PR. Before guys from CN group could make request with `logtail push server`, there maybe two additional PR needed:
> 
> * [x]  Heartbeat `logtail server` address which need update proto buffer definition
> * [x]  Split large `logtail response` if necessary
> 
> But these two additional PR are mainly related with integration test. It wouldn't block CN's development.

**Comment by m-schen** [comment]
mo can run end to end now. 
but sometimes will panic due to log tail order is differ to old pull model.
need to check and fix.
code branch is https://github.com/m-schen/matrixone/tree/cn-push

**Comment by cnutshell** [comment]
> mo can run end to end now. but sometimes will panic due to log tail order is differ to old pull model. need to check and fix. code branch is https://github.com/m-schen/matrixone/tree/cn-push

Try to launch mo-service in my local environment with @m-schen's branch . But there are twists and turns.
After launching `./mo-service -launch ./etc/launch-tae-CN-tae-DN/launch.toml` 11 times, but there is only one successful launching. The failed reasons were `panic` or no response with `mysql` client after a long long running.
I'll continue to make a successful launching for more details. But I'm disappointed about current status, lots of time wasted.

**Comment by cnutshell** [comment]
```bash
./mo-service -launch ./etc/launch-tae-CN-tae-DN/launch.toml >/tmp/dn.log
panic: wait system init timeout

goroutine 1550 [running]:
github.com/matrixorigin/matrixone/pkg/cnservice.(*service).waitSystemInitCompleted(0x140082c2580, {0x1070a7c18, 0x140031dac00})
        /Users/cuiguoke/Github/matrixorigin/matrixone/pkg/cnservice/server_task.go:180 +0x454
github.com/matrixorigin/matrixone/pkg/common/stopper.(*Stopper).doRunCancelableTask.func1()
        /Users/cuiguoke/Github/matrixorigin/matrixone/pkg/common/stopper/stopper.go:259 +0x6c
created by github.com/matrixorigin/matrixone/pkg/common/stopper.(*Stopper).doRunCancelableTask
        /Users/cuiguoke/Github/matrixorigin/matrixone/pkg/common/stopper/stopper.go:254 +0x150
```

Timeout when launching `mo-service`.

**Comment by cnutshell** [comment]
Now, `log tail push model` server publishes incremental log tail in a `batch-based` manner.
and in PR #7718 we have a plan to switch this batch-based model as pure-event-driven model.

Before PR #7718 merged, there are 2 issues from CN side to be resolved:
1. CN consumes log tail slowly now, it couldn't catch up the producer of log tail;
2. CN need to handle the error condition properly when `morpc stream` was closed;

As far as I know, `push model` has been decided to postpone to 0.8.0.

**Comment by fengttt** [comment]
This has been there for too long.   S-1 now.   Must make it work before we can test anything

**Comment by m-schen** [comment]
I have make a pr today. Add a turn on / off flag to use push model.

**Comment by m-schen** [comment]
pr has merged. can set `turn-on-push-model = true` to open the push model.
and add ut for push today.

**Comment by m-schen** [comment]
I am trying to rewrite the code to consume received logs concurrently. but can not start the mo normally now.
maybe tonight or tomorrow morning can ok.
hope every thing goes well.

**Comment by m-schen** [comment]
今日优化了多协程消费log tail的代码，其余进度不大，详情代码可以看我的分支 https://github.com/m-schen/matrixone/blob/cn-push-new/pkg/vm/engine/disttae/logtailPushModel2.go

**Comment by m-schen** [comment]
cn采取并发消费和lazy处理check point后，消费速度得到了提升。目前的性能瓶颈与dn的push logtail间隔不稳定相关，两个response的收取时间经常出现大于设定值的情况，导致query开启时等待时间延长。

**Comment by cnutshell** [comment]
![wecom-temp-346315-70712a212d99c1a3a9b18ce577e09332](https://user-images.githubusercontent.com/20291742/222739343-87c40598-a95f-4592-b28d-fb24913c96b5.jpg)
CN guys noticed that there are some interval spikes when receiving response from DN service.
We've traced the response sending logic of DN, and have noticed that it isn't stable when fetching log tail.
Still work on this issue in order to make the root cause clear.

**Comment by cnutshell** [comment]
> CN guys noticed that there are some interval spikes when receiving response from DN service. We've traced the response sending logic of DN, and have noticed that it isn't stable when fetching log tail. Still work on this issue in order to make the root cause clear.

In PR #8260, the interval spikes should be optimized. In my local environment, the bvt test has improved from 35min to 7min. Detailed information is written in the 8260's comment.

**Comment by m-schen** [comment]
多cn跑bvt不稳定（一次通过81%，一次通过99%）的问题，还没找到原因

**Comment by m-schen** [comment]
https://github.com/m-schen/matrixone/actions/runs/4363189587/jobs/7628983238
将间隔改成20ms后可以通过docker compose的bvt，另外如果是单cn将间隔改得特别小（200微秒）也有概率出现查询结果不对的情况，按代码逻辑应该是出现查询慢的情况，这不符合预期。

**Comment by cnutshell** [comment]
I've make a test tool which could subscribe 3 special tables. By checking timestamp for received logtail responses, this tool could assert that the timestamps are monotonous.
I simulated multi cn under push model via this test tool on my local environment, but failed to reproduce the problem.
https://github.com/matrixorigin/subscriber

**Comment by m-schen** [comment]
fix a lazy load bug may cause query result wrong today.

**Comment by m-schen** [comment]
单cn的docker也跑不起来，存在logtail漏掉或者迟到的现象（这个需要我进一步确认
怀疑是cn与dn所使用的时间戳不同步的问题。

**Comment by cnutshell** [comment]
@aptend has a PR to resolve the consistent issue.

**Comment by m-schen** [comment]
there is  still a unsubscribe work in cn should do.
but I think we can verify the push model now.

**Comment by fengttt** [comment]
Done.   Remaining problems should have their own issue.
